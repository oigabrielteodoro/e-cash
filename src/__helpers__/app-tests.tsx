import React, { useEffect, ReactElement, ReactNode } from 'react'
import { render as rtlRender } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {
  DefaultOptions,
  QueryCache,
  QueryClient,
  QueryClientProvider,
} from 'react-query'
import { MemoryRouter, Routes, Route } from 'react-router-dom'
import { generateId } from 'lib'
import nock from 'nock'

import { pipe } from 'fp-ts/function'
import { TaskEither, map, mapLeft } from 'fp-ts/TaskEither'

import { AppLayout, NotificationContainer } from 'ui'
import { clearToken, queryConfigDefault, setToken } from 'client'

type Options = {
  initialRoute?: string
  initialRoutePath?: string
  authenticated?: boolean
  queryClient?: QueryClient
  routePaths?: string[]
}

type Props = {
  children?: ReactNode
  client?: QueryClient
}

type Callback = (result: unknown) => unknown
type MapAll = (
  fn: Callback,
) => (data: TaskEither<unknown, unknown>) => TaskEither<unknown, unknown>

const queryCache = new QueryCache()
const defaultQueryClient = new QueryClient({
  defaultOptions: queryConfigDefault as DefaultOptions,
  queryCache: queryCache,
})

export function ReactQueryWrapper({
  children,
  client = defaultQueryClient,
}: Props) {
  return <QueryClientProvider client={client}>{children}</QueryClientProvider>
}

export function ReactRouterWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
}: Options & Pick<Props, 'children'>) {
  function MockedPage({ title }: { title: string }) {
    if (authenticated) {
      return (
        <AppLayout>
          <AppLayout.Content>Page {title}</AppLayout.Content>
        </AppLayout>
      )
    }

    return <h1>Page {title}</h1>
  }

  return (
    <MemoryRouter initialEntries={[initialRoute]}>
      <Routes>
        <Route path={initialRoute} element={children} />
        {routePaths?.map((routePath) => (
          <Route
            key={routePath}
            path={routePath}
            element={<MockedPage title={routePath} />}
          />
        ))}
      </Routes>
    </MemoryRouter>
  )
}

export function ReactHookWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
  client = defaultQueryClient,
}: Options & Props) {
  return (
    <ReactQueryWrapper client={client}>
      <ReactRouterWrapper
        initialRoute={initialRoute}
        routePaths={routePaths}
        authenticated={authenticated}
      >
        {children}
      </ReactRouterWrapper>
    </ReactQueryWrapper>
  )
}

function render(
  ui: ReactElement,
  {
    initialRoute = '/',
    queryClient = defaultQueryClient,
    authenticated = true,
    routePaths,
  }: Options = {},
) {
  function Wrapper({ children }: Props) {
    useEffect(() => {
      if (authenticated) {
        setToken(generateId())

        return () => {
          clearToken()
        }
      }
    }, [])

    return (
      <ReactQueryWrapper client={queryClient}>
        <ReactRouterWrapper
          initialRoute={initialRoute}
          routePaths={routePaths}
          authenticated={authenticated}
        >
          {children}
        </ReactRouterWrapper>

        <NotificationContainer />
      </ReactQueryWrapper>
    )
  }

  return {
    ...rtlRender(ui, {
      wrapper: Wrapper,
    }),
  }
}

export function unsafe<T>(value: unknown) {
  return value as T
}

export const mapAll: MapAll = (fn) => (data) => {
  return pipe(data, map(fn), mapLeft(fn))
}

export function getErrorMessage(errors: unknown): string {
  return Array.isArray(errors) ? errors[0].message : ''
}

export * from '@testing-library/react'
export {
  render,
  userEvent,
  nock,
  defaultQueryClient as queryClient,
  queryCache,
}
