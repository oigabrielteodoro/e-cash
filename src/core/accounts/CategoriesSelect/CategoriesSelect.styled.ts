import { theme } from 'config'
import styled from 'styled-components'

export const CategoryOption = styled.div`
  display: flex;
  align-items: center;
`

export const IconContainer = styled.div`
  background: ${theme.colors.blue[100]};
  border-radius: 0.5rem;
  height: 3.2rem;
  width: 3.2rem;
  margin-right: 1rem;
  display: grid;
  place-items: center;

  svg {
    color: ${theme.colors.blue[500]};
  }
`
