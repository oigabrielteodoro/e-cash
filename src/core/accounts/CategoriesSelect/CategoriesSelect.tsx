import React from 'react'
import { AiOutlineBank, AiOutlineBarChart } from 'react-icons/ai'
import { FiDollarSign } from 'react-icons/fi'
import { MdOutlineSavings, MdPayment } from 'react-icons/md'

import { Select, SelectProps } from 'ui'

import * as S from './CategoriesSelect.styled'

export function CategoriesSelect(props: Omit<SelectProps, 'children'>) {
  return (
    <Select renderOptionElementWhenIsSelected {...props}>
      <Select.Option displayValue='Investments' value='investments'>
        <S.CategoryOption>
          <S.IconContainer>
            <AiOutlineBarChart size={20} />
          </S.IconContainer>
          Investments
        </S.CategoryOption>
      </Select.Option>
      <Select.Option displayValue='Savings' value='savings'>
        <S.CategoryOption>
          <S.IconContainer>
            <MdOutlineSavings size={20} />
          </S.IconContainer>
          Savings
        </S.CategoryOption>
      </Select.Option>
      <Select.Option displayValue='Money' value='money'>
        <S.CategoryOption>
          <S.IconContainer>
            <FiDollarSign size={20} />
          </S.IconContainer>
          Money
        </S.CategoryOption>
      </Select.Option>
      <Select.Option displayValue='Checking Account' value='checking_account'>
        <S.CategoryOption>
          <S.IconContainer>
            <AiOutlineBank size={20} />
          </S.IconContainer>
          Checking Account
        </S.CategoryOption>
      </Select.Option>
      <Select.Option displayValue='Payment Account' value='payment_account'>
        <S.CategoryOption>
          <S.IconContainer>
            <MdPayment size={20} />
          </S.IconContainer>
          Payment Account
        </S.CategoryOption>
      </Select.Option>
    </Select>
  )
}
