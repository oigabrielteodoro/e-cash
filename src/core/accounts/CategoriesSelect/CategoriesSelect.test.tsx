import React from 'react'
import { FormProvider, useForm } from 'react-hook-form'

import { render, screen, userEvent } from '__helpers__/app-tests'

import { CategoriesSelect } from './CategoriesSelect'

function MockedComponent() {
  const form = useForm()

  return (
    <FormProvider {...form}>
      <CategoriesSelect name='category' label='Category' />
    </FormProvider>
  )
}

describe('CategoriesSelect', () => {
  it('should be able match snapshot', () => {
    const { container } = render(<MockedComponent />)

    expect(container).toMatchSnapshot()
  })

  it('should be able render correctly', async () => {
    render(<MockedComponent />)

    userEvent.click(
      screen.getByRole('combobox', {
        name: 'category',
      }),
    )

    expect(await screen.findByLabelText('Investments')).toBeInTheDocument()
    expect(await screen.findByLabelText('Savings')).toBeInTheDocument()
    expect(await screen.findByLabelText('Money')).toBeInTheDocument()
    expect(await screen.findByLabelText('Checking Account')).toBeInTheDocument()
    expect(await screen.findByLabelText('Payment Account')).toBeInTheDocument()
  })
})
