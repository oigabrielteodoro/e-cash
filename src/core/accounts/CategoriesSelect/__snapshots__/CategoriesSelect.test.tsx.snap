// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoriesSelect should be able match snapshot 1`] = `
.c1 {
  border: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  position: relative;
}

.c1 label {
  font-size: 1.6rem;
  color: #a1a7ae;
  font-weight: 500;
  cursor: pointer;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0.1rem solid #a1a7ae;
  border-radius: 0.5rem;
  margin-top: 0.8rem;
  -webkit-transition: 300ms;
  transition: 300ms;
  width: 100%;
  background: #ffffff;
}

.c2 > svg,
.c2 .icon {
  margin-left: auto;
  margin-right: 2rem;
  color: #a1a7ae;
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c2:hover {
  border: 0.1rem solid #3a67e1;
}

.c2:hover > svg,
.c2:hover .icon {
  color: #3a67e1;
}

.c2 input,
.c2 textarea {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  border: 0;
  background: transparent;
  padding: 1.8rem 2rem;
  color: #18171d;
  font-size: 1.6rem;
  border-radius: 0.5rem;
}

.c2 input::-webkit-input-placeholder,
.c2 textarea::-webkit-input-placeholder {
  color: #a1a7ae;
}

.c2 input::-moz-placeholder,
.c2 textarea::-moz-placeholder {
  color: #a1a7ae;
}

.c2 input:-ms-input-placeholder,
.c2 textarea:-ms-input-placeholder {
  color: #a1a7ae;
}

.c2 input::placeholder,
.c2 textarea::placeholder {
  color: #a1a7ae;
}

.c2 textarea {
  resize: none;
  padding: 0 2rem;
  margin: 2.4rem 0;
}

.c0 {
  border: 0;
  position: relative;
}

.c3 {
  height: 5.65rem;
}

<div>
  <fieldset
    class="c0"
  >
    <div
      class="c1"
    >
      <label
        for="category"
      >
        Category
      </label>
      <div
        class="c2 c3"
      >
        <input
          aria-label="category"
          autocomplete="off"
          id="category"
          name="category"
          role="combobox"
          value=""
        />
        <svg
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="6 9 12 15 18 9"
          />
        </svg>
      </div>
    </div>
  </fieldset>
</div>
`;
